src/components/JsonInput.vue

```diff
@@ -28,33 +28,55 @@
   import { useScrollLinking } from '@/compositions'
   import { stringify } from '@/utilities/json'
 
+  type ModelValue = string | number | boolean | undefined | null
+  type ExtendedModelValue = ModelValue | ModelValue[] | Record<string, ModelValue>
+
   const props = defineProps<{
-    modelValue: string | undefined,
+    modelValue: ExtendedModelValue,
     showFormatButton?: boolean,
   }>()
 
   const emit = defineEmits<{
-    (event: 'update:modelValue', value: string): void,
+    (event: 'update:modelValue', value: ExtendedModelValue): void,
   }>()
 
+
   const { source, target } = useScrollLinking()
 
   const internalValue = computed({
     get() {
-      return props.modelValue ?? ''
+      return props.modelValue
     },
-    set(val: string) {
+    set(val: ExtendedModelValue) {
       emit('update:modelValue', val)
     },
   })
 
   const format = (): void => {
     try {
-      internalValue.value = stringify(JSON.parse(internalValue.value))
+      internalValue.value = JSON.parse(stringify(internalValue.value))
     } catch {
       // do nothing
     }
   }
+
+  const stringifyInternalValue = computed({
+    get() {
+      return stringify(internalValue.value)
+    },
+    set(val: string | undefined) {
+      try {
+        if (!val) {
+          internalValue.value = undefined
+          return
+        }
+
+        internalValue.value = JSON.parse(val)
+      } catch {
+        // do nothing
+      }
+    },
+  })
```

Comment:
In my experience doing json stringify/parse on a computed like this means that the value you're typing gets formatted as you type causing the cursor to lose its place. There isn't a demo for this component so I cannot verify.